### PR TITLE
Avoid explicit dependsOn(GenerateProtoTask)

### PR DESCRIPTION
**Pull request by ejona86**
The FileCollection can propagate the dependency information implicitly.

### CODE REVIEW

**Comment by ejona86** [comment]
With this change, we should be able to remove the code that changes compileJava and compileKotlin and instead leverage srcSet. That _had_ been failing for me earlier with an infinite loop with a test project, but right now it appears the tests (which is a different environment) are working. (Either I tested it wrong, the environment mattered, or one of the other changes ends up breaking things.) But this is the sort of thing we'll do after this PR:

```diff
diff --git a/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy b/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy
index 713b889..9663cf1 100644
--- a/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy
+++ b/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy
@@ -66,11 +66,6 @@ class ProtobufPlugin implements Plugin<Project> {
             'android-library',
     ]
 
-    private static final List<String> SUPPORTED_LANGUAGES = [
-        'java',
-        'kotlin',
-    ]
-
     private final FileResolver fileResolver
     private Project project
     private ProtobufExtension protobufExtension
@@ -267,9 +262,7 @@ class ProtobufPlugin implements Plugin<Project> {
       }
       processResourcesTask.dependsOn(extractTask)
 
-      SUPPORTED_LANGUAGES.each { String lang ->
-        linkGenerateProtoTasksToTaskName(sourceSet.getCompileTaskName(lang), generateProtoTask)
-      }
+      sourceSet.java.srcDirs generateProtoTask.getOutputSourceDirectorySet()
     }
 
     /**
```

CC @rougsig

**Comment by rougsig** [comment]
Will check a bit, maybe change OutputDirectory to OutputDirectories is the better solution.

https://github.com/google/protobuf-gradle-plugin/pull/585/files#diff-be8d9de6d84a4c949ded2bc324af6067ee934436556e0392188636f473b68777R261

**Comment by ejona86** [comment]
I did try going down the multiple-output-directories route and had trouble extracting only directories from TaskOutputs. To see problems you need to enable `generateDescriptorSet`. I could try that again soon, as I'm more knowledgeable now, but that'd mainly have the benefit of being able to delete `getOutputSourceDirectorySet()`.

**Comment by rougsig** [comment]
Please keep it for me. I'd like to investigate a bit.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy [hunk 1]
```diff
@@ -123,7 +123,6 @@ class ProtobufPlugin implements Plugin<Project> {
     }
 
     private static void linkGenerateProtoTasksToTask(Task task, GenerateProtoTask genProtoTask) {
-      task.dependsOn(genProtoTask)
       task.source genProtoTask.getOutputSourceDirectorySet().include("**/*.java", "**/*.kt")
```
Comment (rougsig) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy:126]: We can pass the task as an argument to the `source` function. dependOn will be added implicitly.

From docs:
[source](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.SourceTask.html#org.gradle.api.tasks.SourceTask:source(java.lang.Object[]))(sources)	
Adds some source to this task. The given source objects will be evaluated as per [Project.files(java.lang.Object[])](https://docs.gradle.org/current/dsl/org.gradle.api.Project.html#org.gradle.api.Project:files(java.lang.Object[])).

[ConfigurableFileCollection](https://docs.gradle.org/current/javadoc/org/gradle/api/file/ConfigurableFileCollection.html) Project.files([Object](https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html)... paths)
You can pass any of the following types to this method:
<...>
A [Task](https://docs.gradle.org/current/dsl/org.gradle.api.Task.html). Converted to the task's output files. The task is executed if the file collection is used as an input to another task.

**Review by YifeiZhuang — APPROVED** [review]

**Pull request merged**